### DISCUSSION TITLE
Hot Reload with Custom Configuration not working

### DISCUSSION THREAD

**Discussion by phillippschmedt** [dotnet/maui#22163] [Q&A]
Hi,

I am using custom configuration like this. 
        `<Configurations>DevDebug;NeutralDevDebug;DevRelease;NeutralDevRelease;Prod;NeutralProd</Configurations>`
I somehow managed to set all the build properties correctly to make iOS/Android debug and release builds work in general. Unfortunately I can't seem to get hot reload of xaml to work for iOS or Android. 

Can anyone help out here? 

Cheers,
Phillipp

**Comment by phillippschmedt** [comment]
Also I am aware of this discussion and the discussion linked: https://github.com/dotnet/maui/discussions/21678 
I am looking for a solution that work's now.

**Comment by LeoJHarris** [comment]
Not a true fix but a work around, you coulld a Debug Configuration for the purpose of hot reload and make sure it shares the same config setting in .csproj as the custom configuration.

**Comment by sergeevgk** [comment]
Unfortunately, according to the documentation there is no workaround https://learn.microsoft.com/en-us/dotnet/maui/xaml/hot-reload?view=net-maui-10.0#troubleshooting. (same for MAUI 8 and 9)
> XAML Hot Reload requires your debug configuration to be named Debug. If a custom build configuration is used, XAML Hot Reload will stop working if the build configuration isn't named Debug.

Tried everything but only made so far to enable the XAML Live Tree Preview when specifying `<EnableHotReload>true</EnableHotReload>` in the custom configuration.

**Reply by LeoJHarris** [reply]
Not sure if it helps you but its how we do it, create `Debug` configuration, then within the `csproj` file you can do something like:

```csproj
<PropertyGroup>
<TargetFrameworks>net9.0-android;net9.0-ios</TargetFrameworks>
<Configurations>Debug;Foo1;Foo2;</Configurations>
</PropertyGroup>

<PropertyGroup Condition="'$(Configuration)' == 'Debug'">
	<DefineConstants>DEBUG;FOO1</DefineConstants>
</PropertyGroup>
<PropertyGroup Condition="'$(Configuration)'=='Foo1Debug'">
	<DefineConstants>FOO1;DEBUG;</DefineConstants>
</PropertyGroup>
<PropertyGroup Condition="'$(Configuration)'=='Foo2Debug'">
	<DefineConstants>FOO2;DEBUG;</DefineConstants>
</PropertyGroup>

<Choose>
	<When Condition="$(DefineConstants.Contains('DEBUG'))">
		
	</When>
	<Otherwise>
		
	</Otherwise>
</Choose>

<Choose>
	<When Condition="$(DefineConstants.Contains('FOO1'))">

         </When>
	<Otherwise>
		
	</Otherwise>
</Choose>
```

When you need hot reload to work switch to `Debug` Configuration and just make sure the csproj file applies what you need for which ever custom config you opted for.

Furthermore when publishing just provide the -c flag for the specific custom release config like so:

`dotnet publish -f net9.0-android --configuration FooRelease`

**Reply by AlleSchonWeg** [reply]
Be careful with custom configurations names. Internally MAUI/.NET set some MSBuild properties depending on the names "Release" and "Debug". If you have a name like DevRelease and expect a Release-Build you will not get a Release-Build. 
https://github.com/dotnet/sdk/issues/31918

**Reply by MerSpyro** [reply]
Exactly as @LeoJHarris says. Create a custom property in csproj and setvalues depending on that.
@davidortinau 
Is it possible to add this to the documentation, as it's not 1st or 2nd time people are asking how to make custom configurations?

**Reply by LeoJHarris** [reply]
> Be careful with custom configurations names. Internally MAUI/.NET set some MSBuild properties depending on the names "Release" and "Debug". If you have a name like DevRelease and expect a Release-Build you will not get a Release-Build. [dotnet/sdk#31918](https://github.com/dotnet/sdk/issues/31918)

@AlleSchonWeg We are aware of it but what's the alternative? We have been doing this since the Xamarin days and we have not noticed any issues. That being said if you know the complete list of MSBuild Properties you can manually add them to csproj anyway.

**Reply by AlleSchonWeg** [reply]
> > Be careful with custom configurations names. Internally MAUI/.NET set some MSBuild properties depending on the names "Release" and "Debug". If you have a name like DevRelease and expect a Release-Build you will not get a Release-Build. [dotnet/sdk#31918](https://github.com/dotnet/sdk/issues/31918)
> 
> @AlleSchonWeg We are aware of it but what's the alternative? We have been doing this since the Xamarin days and we have not noticed any issues. That being said if you know the complete list of MSBuild Properties you can manually add them to csproj anyway.

There is no perfect alternative. I tested a lot and never had all properties set like the default debug/release configurations. Finally i created a custom MSBuild property `MyEnv` and set it at the beginning of my csproj and comment/uncomment what i need. By default `MyEnv` is not set, when i build the project with VS:
```XML
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup Condition="'$(Configuration)'=='Debug' And '$(MyEnv)'==''">
		<!--<MyEnv>LocalDebug</MyEnv>-->
	</PropertyGroup>

	<PropertyGroup Condition="'$(Configuration)'=='Release' And '$(MyEnv)'==''">
		<!--<MyEnv>Prod</MyEnv>-->
		<!--<MyEnv>Staging1</MyEnv>-->
		<!--<MyEnv>Staging2</MyEnv>-->
	</PropertyGroup>
	...
</Project>
```
Then i set values depending on `MyEnv`:
```XML
	<PropertyGroup>
		<DefineConstants Condition="'$(MyEnv)'=='' And '$(Configuration)'=='Release'">$(DefineConstants);PROD;XX;YY</DefineConstants>
		<DefineConstants Condition="'$(MyEnv)'=='' And '$(Configuration)'=='Debug'">$(DefineConstants);LOCAL;XX;ZZ</DefineConstants>
		<DefineConstants Condition="'$(MyEnv)'=='Staging1'">$(DefineConstants);INT1;XX;YY</DefineConstants>
		<DefineConstants Condition="'$(MyEnv)'=='Staging2'">$(DefineConstants);INT2;XX;YY</DefineConstants>
	</PropertyGroup>
```
You can also set other properties like outputpath depending on `MyEnv`. When build/publish with cli and you want Staging1 for example you can set `MyEnv` to Staging1 (`dotnet build /p:MyEnv=Staging1`). It depends on you what you prefer. Set via csproj or via cli.

**Comment by simeyla** [comment]
I get the same behavior with Blazor when trying to use a custom configuration I called `Watch`.

    dotnet watch run -c Watch 

All the messages in the console seem to indicate it's working and I even get a little animation in my browser when the hot-reload succeeds but it doesn't actually work. Since it's documented as not working in Maui I assume it's the same basic logic. Very odd though. 

Incidentally the reason I wanted it was so I could simultaneously run `dotnet watch` as well as build in Visual Studio without running into `.dll` collisions.
